src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+            #       flag, but we still need some proper logging for these.
+            msg = "Skipping workflow: %s - 'ParentageResolved' flag set to false."
+            msg += "\nWill retry again in the next cycle."
+            self.logger.info(msg, wflow['RequestName'])
+        elif wflow['RequestStatus'] == 'announced' and wflow['ParentageResolved']:
+            for pline in self.cleanuplines:
+                try:
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = "%s: General error from pipeline. Error: \n%s. "
+                    msg += "\nWill retry again in the next cycle."
+                    self.logger.exception(msg, pline.name, str(ex))
+                    continue
+                if wflow['CleanupStatus'][pline.name]:
+                    self.wfCounters['cleaned'][pline.name] += 1
+        else:
+            # We shouldn't be here:
+            msg = "Skipping workflow: %s - "
+            msg += "Does not fall under any of the defined categories."
+            self.logger.info(msg, wflow['RequestName'])
+
+        # Archive:
+        try:
+            self.plineArchive.run(wflow)
+            self.wfCounters['archived'] += 1
+        except MSRuleCleanerArchival as ex:
+            msg = "%s: Archival Error: %s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.error(msg, wflow['PlineMarkers'][-1], ex.message())
+        except Exception as ex:
+            msg = "%s General error from pipeline. Error: \n%s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.exception(msg, wflow['PlineMarkers'][-1], str(ex))
+        return wflow
+
+    def setPlineMarker(self, wflow, pName=None):
+        """
+        A function intended to mark which is the pipeline currently working
+        on the workflow. It is supposed to be called always as a first function
+        in the pipeline.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :param  pName:   The name of the functional pipeline
+        :return wflow:
+        """
+        # NOTE: The current functional pipeline MUST always be appended at the
+        #       end of the 'PlineMarkers' list
+
+        # First get rid of the default:
+        if not wflow['PlineMarkers']:
+            wflow['PlineMarkers'] = []
+
+        # Then push our current value into the markers list:
+        wflow['PlineMarkers'].append(pName)
+
+        # Populate the list of flags to be used later:
+        if pName not in wflow['RulesToClean']:
+            if pName in [pline.name for pline in self.cleanuplines]:
+                wflow['RulesToClean'][pName] = []
+        if pName not in wflow['CleanupStatus']:
+            if pName in [pline.name for pline in self.cleanuplines]:
+                wflow['CleanupStatus'][pName] = False
+        return wflow
+
+    def _checkClean(self, wflow):
```

Comment:
I have a hard time picturing all the combinations here. It's also a method that can screw up with the workflow cleanup. So, I'd suggest to get a unit test exercising this logic. No need to run the whole microservice, but this method is very important IMO.

Reply:
So If you allow me I may try to visualize it like that:
We start from the fact that we apply a mask of named flags from one list (the list of pipelines that have worked on a workflow) over a list of `Bool` flags (The list of cleanup results). And we also need to take into account that the two lists may not overlap because of the fact that not every workflow will fall under the same logic of cleaning - some may have cleaned just from rules from the Agents other from all MStranferoror and Agents and so on. So the table of valid use cases is [1]. When the last one i consider a bug because it will produce wrong results. That's why we  have  the last condition in the code above.




[1]
```
|--------------+---------------------------+---------------------+---------------------------------------|
| Use cases    | Lists                     | Contents            | Example                               |
|--------------+---------------------------+---------------------+---------------------------------------|
| MasList      | Mask:                     | M1  M2  --          | [plMStrCont, plMStrBlock]             |
| shorter than | Flags:                    | F1  F2  F_extraFlag | [True,       True,        False ]     |
| FlagList     | Flags taken into account: | {F1, F2}            |                                       |
|              | Final Value:              | all([F1, F2])       | all([True, True]) = True              |
|              |                           |                     |                                       |
|--------------+---------------------------+---------------------+---------------------------------------|
| MaskList     | Mask:                     | --                  | []                                    |
| Empty        | Flags:                    | F1 F2 F3            | [True,          True,        False ]  |
|              | Flags taken into account: | --                  |                                       |
|              | Final Value:              | False               | False                                 |
|              |                           |                     |                                       |
|--------------+---------------------------+---------------------+---------------------------------------|
| MaskList     | Mask:                     | M1  M2  M3_extra    | [plMStrCont, plMStrBlock, plArchiver] |
| longer than  | Flags:                    | F1  F2              | [True,       True ]                   |
| Flaglist     | Flags taken into account: | {F1, F2}            |                                       |
|              | Final Value:              | all([F1, F2])       | all([True, True]) = True              |
|              |                           |                     |                                       |
|--------------+---------------------------+---------------------+---------------------------------------|
| FlagList     | Mask:                     | M1  M2              | [plMStrCont, plMStrBlock]             |
| empty        | Flags:                    | --  --              | []                                    |
|              | Flags taken into account: | {}                  |                                       |
|              | Final Value:              | all([])             | all([]) = True                        |
|              |                           |                     |                                       |
|--------------+---------------------------+---------------------+---------------------------------------|
```

Reply:
Thanks, let's see how the unit tests look like ;)

Reply:
Done - The unit tests are implemented and are referring this comment for visualization purposes of the test itself.